From: Jeff Mahoney <jeffm@suse.com>
Subject: [PATCH 04/15] ocfs2: complete failure recovery for nodemanager init
Patch-mainline: never - Already in 2.6.16-rc6, but backed out by ocfs2-1.2 (ocfs2-revert-2.6.16-rc5-git7.diff)

 This patch finishes cleaning up the node manager allocations if it fails
 to initialize.

 fs/ocfs2/cluster/nodemanager.c |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletion(-)

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

diff -ruNpX ../dontdiff linux-2.6.15-git12.staging1/fs/ocfs2/cluster/nodemanager.c linux-2.6.15-git12.staging2/fs/ocfs2/cluster/nodemanager.c
--- linux-2.6.15-git12.staging1/fs/ocfs2/cluster/nodemanager.c	2006-01-19 17:52:22.626312392 -0500
+++ linux-2.6.15-git12.staging2/fs/ocfs2/cluster/nodemanager.c	2006-01-19 17:52:22.710299624 -0500
@@ -861,7 +861,7 @@ static int __init init_o2nm(void)
 	if (!ocfs2_table_header) {
 		printk(KERN_ERR "nodemanager: unable to register sysctl\n");
 		ret = -ENOMEM; /* or something. */
-		goto out;
+		goto out_o2net;
 	}
 
 	ret = o2net_register_hb_callbacks();
@@ -894,6 +894,7 @@ static int __init init_o2nm(void)
 	if (ret == 0)
 		goto out;
 
+	o2net_proc_exit(o2nm_proc);
 out_mlog:
 	mlog_remove_proc(o2nm_proc);
 out_remove:
@@ -904,6 +907,8 @@ out_callbacks:
 	o2net_unregister_hb_callbacks();
 out_sysctl:
 	unregister_sysctl_table(ocfs2_table_header);
+out_o2net:
+	o2net_exit();
 out:
 	return ret;
 }
